### PR TITLE
[MODULE] Improvements to module 3 on PEFT

### PR DESCRIPTION
**Pull request by burtenshaw**
This PR finalises module 3 on parameter efficient fine-tuning methods. Set to be released on December 9th.

thanks @Aaryan2304 for starting this. 

## Current Status

- the prose in readmes is finished and just need a copy review
- the 'load lora adapter' exercise notebook needs to have exercise blocks added. see the [sft example](https://github.com/huggingface/smol-course/blob/main/1_instruction_tuning/notebooks/sft_finetuning_example.ipynb)

### CODE REVIEW

**Comment by burtenshaw** [comment]
@Aaryan2304 let me know if you'd like to take on any of the steps above. Happy to collaborate!

**Comment by burtenshaw** [comment]
@alvarobartt @davidberenstein1957 you both offered reviews, so I assigned you here, but no pressure.

**Review by Aaryan2304 — APPROVED** [review]

File: 3_parameter_efficient_finetuning/README.md [hunk 1]
```diff
@@ -11,20 +11,29 @@ Traditional fine-tuning updates all model parameters during training, which beco
...
+### 1️⃣ LoRA (Low-Rank Adaptation)
 
-### [Prompt Tuning](./prompt_tuning.md)
+LoRA has emerged as the most widely adopted PEFT method, offering an elegant solution to efficient model adaptation. Instead of modifying the entire model, LoRA injects trainable matrices into the model's attention layers. This approach typically reduces trainable parameters by about 90% while maintaining comparable performance to full fine-tuning. We will explore LoRA in the [LoRA (Low-Rank Adaptation)](./lora_adapters.md) section.
+ 
+### 2️⃣ Prompt Tuning
 
-Prompt tuning offers an even lighter approach by adding trainable tokens to the input rather than modifying model weights.
+Prompt tuning offers an even lighter approach by adding trainable tokens to the input rather than modifying model weights. Prompt tuning is less popular than LoRA, but can be a useful technique for quickly adapting a model to new tasks or domains. We will explore prompt tuning in the [Prompt Tuning](./prompt_tuning.md) section.
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/README.md:22]:
```suggestion
Prompt tuning offers an **even lighter** approach by **adding trainable tokens to the input** rather than modifying model weights. Prompt tuning is less popular than LoRA, but can be a useful technique for quickly adapting a model to new tasks or domains. We will explore prompt tuning in the [Prompt Tuning](./prompt_tuning.md) section.
```

File: 3_parameter_efficient_finetuning/README.md [hunk 2]
```diff
@@ -11,20 +11,29 @@ Traditional fine-tuning updates all model parameters during training, which beco
...
 ## Available Methods
 
-### [LoRA (Low-Rank Adaptation)](./lora_adapters.md)
+In this module, we will cover two popular PEFT methods:
 
-LoRA has emerged as the most widely adopted PEFT method, offering an elegant solution to efficient model adaptation. Instead of modifying the entire model, LoRA injects trainable rank decomposition matrices into the model's attention layers. This approach typically reduces trainable parameters by about 90% while maintaining comparable performance to full fine-tuning.
+### 1️⃣ LoRA (Low-Rank Adaptation)
 
-### [Prompt Tuning](./prompt_tuning.md)
+LoRA has emerged as the most widely adopted PEFT method, offering an elegant solution to efficient model adaptation. Instead of modifying the entire model, LoRA injects trainable matrices into the model's attention layers. This approach typically reduces trainable parameters by about 90% while maintaining comparable performance to full fine-tuning. We will explore LoRA in the [LoRA (Low-Rank Adaptation)](./lora_adapters.md) section.
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/README.md:18]:
```suggestion
LoRA has emerged as the most widely adopted PEFT method, offering an elegant solution to efficient model adaptation. Instead of modifying the entire model, **LoRA injects trainable matrices into the model's attention layers.** This approach typically reduces trainable parameters by about 90% while maintaining comparable performance to full fine-tuning. We will explore LoRA in the [LoRA (Low-Rank Adaptation)](./lora_adapters.md) section.
```

File: 3_parameter_efficient_finetuning/README.md [hunk 3]
```diff
@@ -11,20 +11,29 @@ Traditional fine-tuning updates all model parameters during training, which beco
...
+| Load LoRA Adapters | Learn how to load and use trained LoRA adapters | 🐢 Load pre-trained adapters<br>🐕 Merge adapters with base model<br>🦁 Switch between multiple adapters | [Notebook](./notebooks/load_lora_adapter_example.ipynb) | <a target="_blank" href="https://colab.research.google.com/github/huggingface/smol-course/blob/main/3_parameter_efficient_finetuning/notebooks/load_lora_adapter_example.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a> |
+<!-- | Prompt Tuning | Learn how to implement prompt tuning | 🐢 Train soft prompts<br>🐕 Compare different initialization strategies<br>🦁 Evaluate on multiple tasks | [Notebook](./notebooks/prompt_tuning_example.ipynb) | <a target="_blank" href="https://colab.research.google.com/github/huggingface/smol-course/blob/main/3_parameter_efficient_finetuning/notebooks/prompt_tuning_example.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a> | -->
 
 ## Resources
 - [PEFT Documentation](https://huggingface.co/docs/peft)
 - [LoRA Paper](https://arxiv.org/abs/2106.09685)
+- [QLoRA Paper](https://arxiv.org/abs/2305.14314)
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/README.md:35]: Are we using this in the course?

Comment (davidberenstein1957) [reply 3_parameter_efficient_finetuning/README.md:35]: If so, perhaps specifically mention them. If not, we can perhaps remove them?

File: 3_parameter_efficient_finetuning/lora_adapters.md [hunk 4]
```diff
@@ -1,31 +1,61 @@
...
+## Understanding LoRA
 
-config = LoraConfig(
-    task_type="CAUSAL_LM",
-    r=8,  # Rank of update matrices
-    lora_alpha=32,  # Scale of updates
-    lora_dropout=0.1
-)
-```
+LoRA (Low-Rank Adaptation) is a parameter-efficient fine-tuning technique that freezes the pre-trained model weights and injects trainable rank decomposition matrices into the model's layers. Instead of training all model parameters during fine-tuning, LoRA decomposes the weight updates into smaller matrices through low-rank decomposition, significantly reducing the number of trainable parameters while maintaining model performance. For example, when applied to GPT-3 175B, LoRA reduced trainable parameters by 10,000x and GPU memory requirements by 3x compared to full fine-tuning. You can read more about LoRA in the [LoRA paper](https://arxiv.org/pdf/2106.09685).
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/lora_adapters.md:7]: perhaps briefly introduce QLoRA as even more efficient model.

File: 3_parameter_efficient_finetuning/lora_adapters.md [hunk 5]
```diff
@@ -1,31 +1,61 @@
...
 
-This configuration can be used with TRL to train a model with LoRA. LoRA matrices are known as adapters and can be merged back into the base model.
+LoRA works by adding pairs of rank decomposition matrices to transformer layers, typically focusing on attention weights. During inference, these adapter weights can be merged with the base model, resulting in no additional latency overhead. LoRA is particularly useful for adapting large language models to specific tasks or domains while keeping resource requirements manageable.
 
 ## Merging LoRA Adapters
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/lora_adapters.md:11]: Isn't a more common scenario of LoRA adapters doing something like multi-adapter serving? Perhaps we should have a brief chapter about that too?

image

File: 3_parameter_efficient_finetuning/lora_adapters.md [hunk 6]
```diff
@@ -1,31 +1,61 @@
...
+The merging process requires attention to memory management and precision. Since you'll need to load both the base model and adapter weights simultaneously, ensure sufficient GPU/CPU memory is available. Using `device_map="auto"` in `transformers` will help with automatic memory management. Maintain consistent precision (e.g., float16) throughout the process, matching the precision used during training and saving the merged model in the same format for deployment. Before deploying, always validate the merged model by comparing its outputs and performance metrics with the adapter-based version.
 
-### Basic Merging Process
+Adapters are also be convenient for switching between different tasks or domains. You can load the base model and adapter weights separately. This allows for quick switching between different task-specific weights.
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/lora_adapters.md:17]: I feel there can be a brief section about this maybe "Using LoRA adapters" or "Inference with LoRA adapters" even before the Merging part.

File: 3_parameter_efficient_finetuning/lora_adapters.md [hunk 7]
```diff
@@ -1,31 +1,61 @@
...
 
-### Basic Merging Process
+Adapters are also be convenient for switching between different tasks or domains. You can load the base model and adapter weights separately. This allows for quick switching between different task-specific weights. 
 
-Here's how to merge a LoRA adapter back into the base model:
+## Implementation Guide
+
+The `notebooks/` directory contains practical tutorials and exercises for implementing different PEFT methods. Begin with `load_lora_adapter_example.ipynb` for a basic introduction, then explore `lora_finetuning.ipynb` for a more detailed look at how to fine-tune a model with LoRA and SFT.
+
+When implementing PEFT methods, start with small rank values (4-8) for LoRA and monitor training loss. Use validation sets to prevent overfitting and compare results with full fine-tuning baselines when possible. The effectiveness of different methods can vary by task, so experimentation is key.
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/lora_adapters.md:23]: perhaps briefly mention it is less efficient than full-fine-tuning but it is good to explore and iterate to set expectations.

File: 3_parameter_efficient_finetuning/lora_adapters.md [hunk 8]
```diff
@@ -1,31 +1,61 @@
...
+)
+
+# Load model on specific device
+model = AutoModelForCausalLM.from_pretrained(
+    "your-model-name",
+    load_in_8bit=True,  # Optional: use 8-bit precision
+    device_map="auto",
+    peft_config=lora_config
+)
+```
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/lora_adapters.md:49]: in this example, we don't mention or use TRL, but it is in the TRL section. What do you want to showcase here?

File: 3_parameter_efficient_finetuning/lora_adapters.md [hunk 9]
```diff
@@ -1,31 +1,61 @@
...
-Here's how to merge a LoRA adapter back into the base model:
+## Implementation Guide
+
+The `notebooks/` directory contains practical tutorials and exercises for implementing different PEFT methods. Begin with `load_lora_adapter_example.ipynb` for a basic introduction, then explore `lora_finetuning.ipynb` for a more detailed look at how to fine-tune a model with LoRA and SFT.
+
+When implementing PEFT methods, start with small rank values (4-8) for LoRA and monitor training loss. Use validation sets to prevent overfitting and compare results with full fine-tuning baselines when possible. The effectiveness of different methods can vary by task, so experimentation is key.
+
+## Using TRL with PEFT
+
+PEFT methods can be combined with TRL (Transformers Reinforcement Learning) for efficient fine-tuning. This integration is particularly useful for RLHF (Reinforcement Learning from Human Feedback) as it reduces memory requirements.
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/lora_adapters.md:27]:
```suggestion
PEFT methods can be combined with TRL (Transformers Reinforcement Learning) for efficient fine-tuning. This integration is particularly useful for RLHF (Reinforcement Learning from Human Feedback) techniques like DPO and ORPO as it reduces memory requirements.
```

File: 3_parameter_efficient_finetuning/lora_adapters.md [hunk 10]
```diff
@@ -44,7 +74,11 @@ peft_model = PeftModel.from_pretrained(
 )
 
 # 3. Merge adapter weights with base model
-merged_model = peft_model.merge_and_unload()
+try:
+    merged_model = peft_model.merge_and_unload()
+except RuntimeError as e:
+    print(f"Merging failed: {e}")
+    # Implement fallback strategy or memory optimization
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/lora_adapters.md:81]: Perhaps the try-except is not needed because we already get an error message from the SDK so we can just focus on the happy flow.

File: 3_parameter_efficient_finetuning/lora_adapters.md [hunk 11]
```diff
@@ -1,31 +1,61 @@
...
 
-This configuration can be used with TRL to train a model with LoRA. LoRA matrices are known as adapters and can be merged back into the base model.
+LoRA works by adding pairs of rank decomposition matrices to transformer layers, typically focusing on attention weights. During inference, these adapter weights can be merged with the base model, resulting in no additional latency overhead. LoRA is particularly useful for adapting large language models to specific tasks or domains while keeping resource requirements manageable.
 
 ## Merging LoRA Adapters
 
 After training with LoRA, you might want to merge the adapter weights back into the base model for easier deployment. This creates a single model with the combined weights, eliminating the need to load adapters separately during inference.
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/lora_adapters.md:13]: I would first introduce code to simply load one or multiple adapters instead of focusing on the merging already. Are there clear benefits of merging besides the mentioned "easier than loading separately"? If so, can you mention them?

File: 3_parameter_efficient_finetuning/prompt_tuning.md [hunk 12]
```diff
@@ -31,12 +39,6 @@ peft_config = PromptTuningConfig(
 model = get_peft_model(model, peft_config)
 ```
 
-## Key Components
-
-The core of prompt tuning revolves around virtual tokens, which are trainable embeddings added to the input. These typically number between 8 and 32 tokens and can be initialized either randomly or from existing text. The initialization method plays a crucial role in the training process, with text-based initialization often performing better than random initialization.
-
-During training, only the prompt parameters are updated while the base model remains frozen. This focused approach uses standard training objectives but requires careful attention to the learning rate and gradient behavior of the prompt tokens.
-
 ## Comparison with Other Methods
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/prompt_tuning.md:42]:
```suggestion
## Comparison to Other Methods
```

File: 3_parameter_efficient_finetuning/prompt_tuning.md [hunk 13]
```diff
@@ -31,12 +39,6 @@ peft_config = PromptTuningConfig(
...
 
 When compared to other PEFT approaches, prompt tuning stands out for its efficiency. While LoRA offers low parameter counts and memory usage but requires loading adapters for task switching, prompt tuning achieves even lower resource usage and enables immediate task switching. Full fine-tuning, in contrast, demands significant resources and requires separate model copies for different tasks.
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/prompt_tuning.md:44]: perhaps this section is better placed in the overal introduction section?

File: 3_parameter_efficient_finetuning/lora_adapters.md [hunk 14]
```diff
@@ -59,56 +93,13 @@ merged_model.save_pretrained("path/to/save/merged_model")
...
-# Load model on specific device
-model = AutoModelForCausalLM.from_pretrained(
-    "your-model-name",
-    load_in_8bit=True,  # Optional: use 8-bit precision
-    device_map={"": current_device},  # Assign to correct device
-    peft_config=lora_config
-)
-```
+⏩ Move on to the [Prompt Tuning](prompt_tuning.md) guide to learn how to fine-tune a model with prompt tuning.
+⏩ Move on the [Load LoRA Adapters Tutorial](./notebooks/load_lora_adapter.ipynb) to learn how to load LoRA adapters.
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/lora_adapters.md:99]: Why are there 2 options here? Don't we want to have a single flow through the course?

File: 3_parameter_efficient_finetuning/prompt_tuning.md [hunk 15]
```diff
@@ -51,15 +53,22 @@ When implementing prompt tuning, start with a small number of virtual tokens (8-
...
+2. Resource-constrained environments
+3. Rapid task adaptation
+4. Privacy-sensitive applications
 
-Prompt tuning excels in several scenarios, particularly for classification tasks and simple generation tasks. Its minimal resource requirements make it ideal for environments with limited computational resources. The ability to quickly switch between tasks also makes it valuable for multi-task applications where different behaviors are needed at different times.
+As models get smaller, prompt tuning becomes less competitive compared to full fine-tuning. For example, on models like SmolLM2 scales prompt tuning is less relevant than full fine-tuning. 
 
 ## Next Steps
 
-To get hands-on experience with prompt tuning, try the [Prompt Tuning Tutorial](./notebooks/prompt_tuning_example.ipynb). This practical guide will walk you through implementing the technique with your own model and data.
```
Comment (davidberenstein1957) [inline 3_parameter_efficient_finetuning/prompt_tuning.md:60]: is there not going to be a promt_tuning_example anymore?

**Review by davidberenstein1957 — COMMENTED** [review]
Looking great :) Some small nits w.r.t. typo's and phrasing. Also, I might be wrong but I feel people normally load and serve as easiest LoRA option and after that consider merging or serving multiple adapters? I might be wrong though but if you agree that flow can be improved a bit IMO.

File: 3_parameter_efficient_finetuning/prompt_tuning.md (hunk 13 above)
Comment (davidberenstein1957) [reply 3_parameter_efficient_finetuning/prompt_tuning.md:44]: Specifically the table is a nice overview to understand what is coming.

File: 3_parameter_efficient_finetuning/lora_adapters.md (hunk 5 above)
Comment (burtenshaw) [reply 3_parameter_efficient_finetuning/lora_adapters.md:11]: That's a nice suggestion. Feel free to open a PR on it.

File: 3_parameter_efficient_finetuning/lora_adapters.md (hunk 11 above)
Comment (burtenshaw) [reply 3_parameter_efficient_finetuning/lora_adapters.md:13]: At smaller model sizes like this, merging is more likely. But you're right, we could expand this for educational purposes. I've added this issue: https://github.com/huggingface/smol-course/issues/72#issue-2727951665

File: 3_parameter_efficient_finetuning/lora_adapters.md (hunk 14 above)
Comment (burtenshaw) [reply 3_parameter_efficient_finetuning/lora_adapters.md:99]: The course does this a lot to allow people to move around at different paces, but I can how it's confusing.

File: 3_parameter_efficient_finetuning/prompt_tuning.md (hunk 15 above)
Comment (burtenshaw) [reply 3_parameter_efficient_finetuning/prompt_tuning.md:60]: Nope 😢
